builder/proxmox/common/builder.go

```diff
@@ -56,37 +49,21 @@ func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (pack
 	}
 
 	// Set up the state
-	state := new(multistep.BasicStateBag)
 	state.Put("config", &b.config)
 	state.Put("proxmoxClient", b.proxmoxClient)
 	state.Put("hook", hook)
 	state.Put("ui", ui)
 
-	// Build the steps
-	steps := []multistep.Step{
-		&common.StepDownload{
-			Checksum:    b.config.ISOChecksum,
-			Description: "ISO",
-			Extension:   b.config.TargetExtension,
-			ResultKey:   downloadPathKey,
-			TargetPath:  b.config.TargetPath,
-			Url:         b.config.ISOUrls,
-		}}
-
-	for idx := range b.config.AdditionalISOFiles {
-		steps = append(steps, &common.StepDownload{
-			Checksum:    b.config.AdditionalISOFiles[idx].ISOChecksum,
-			Description: "additional ISO",
-			Extension:   b.config.AdditionalISOFiles[idx].TargetExtension,
-			ResultKey:   b.config.AdditionalISOFiles[idx].downloadPathKey,
-			TargetPath:  b.config.AdditionalISOFiles[idx].downloadPathKey,
-			Url:         b.config.AdditionalISOFiles[idx].ISOUrls,
-		})
+	comm := &b.config.Comm
+	if state.Get("comm") != nil {
+		comm = state.Get("comm").(*communicator.Config)
```

Comment:
Isn't `state.Get("comm")` always the same object as `&b.config.Comm` when it is set (ie when building with the clone builder)?

Reply:
this is the tricky part - we have two "builder" references here, one that's the generic one (called here), and one that's the implementation-specific one (iso or clone, calling the generic one).

&b here always refers to the generic common builder ref, so I'm passing the communicator through the state as that's shared between the builders.

The clone setup ssh connection config don't have &b.config reference and cannot access/update that one, only the clone builder ref. Using &b.config.Comm directly, I was not able to get a successful SSH connection without this bit.

Reply:
Hm, but the specific ones both embed the common config, not duplicate it, so there should only be one instance of `comm`? If you check the pointers for `b.config.Comm` and `state.Get("comm")`, are they different? That would be worrying, if so.